### PR TITLE
feat(#155): v2 lifecycle reconciler

### PR DESCRIPTION
**Pull request by evanstern**
## Summary

Adds the reconciler layer that detects stale lifecycle rows and transitions them to a terminal state before downstream consumers (e.g. #149's message bus) read from the state store. Closes #155.

Before this card, #148's lifecycle writes \`state=running\` with a pid and tmux_session but nothing verifies those processes still exist. If a user \`tmux kill-session\`s the orchestrator or opencode crashes, the row stays \`running\` forever, and #149's message bus would queue messages against tombstoned rows with no observable failure. The reconciler closes that gap.

## What's in the box

- **Schema migration \`002_reconciler.sql\`** — extends \`orchestrators.state\` CHECK to include \`stale\` via the table-swap pattern (SQLite can't alter CHECK constraints in place) and adds \`stale_reason TEXT\` to both \`orchestrators\` and \`features\`. Second exercise of the #156 migration infrastructure. Schema version bumps from 1 to 2.
- **\`internal/lifecycle/reconciler.go\`** — \`Manager.Reconcile(ctx, name)\`, a public \`Prober\` interface, and a default tmux/pid prober. Orchestrators in \`starting|running|stopping\` are probed against \`tmux has-session\` and pid liveness (Linux \`/proc\` with signal-0 fallback, zombie-aware). Features in \`spawning|running|reporting\` are probed against their tmux session only (session_id liveness is deferred to #149 per brief §5.1). A 30-second freshness window protects in-flight \`StartOrchestrator\` from being false-marked.
- **\`StartOrchestrator\` guard widened** — now accepts \`state IN ('stopped','stale')\` and clears \`stale_reason\` on forward transition, so operators restart a stale orch without hand-editing the DB.
- **Hook event \`post-orchestrator-stale\`** — fires after each orchestrator → stale transition, payload follows existing orchestrator hook schema with \`stale_reason\` included.
- **CLI: \`coda-core orchestrator reconcile [<name>] [--json]\`** — explicit reconcile command. Exit codes 0/1/2; reconcile never emits exit 3 (not a lifecycle-blocked event, per brief §3.6).
- **Lazy reconcile** on \`orchestrator ls\` and \`status\` unless \`CODA_NO_AUTO_RECONCILE=1\`. Silent on this path — no stderr noise per brief §5.8.
- **Tests** — \`internal/lifecycle/reconciler_test.go\` covers tmux-dead, pid-dead, both-alive, freshness-window, idempotent-stale, feature-failed, hook-fires (spy dispatcher pattern from \`ordering_test.go\`), and start-from-stale using a stub \`Prober\`. \`test/v2-lifecycle.bats\` extended with reconcile CLI, \`CODA_NO_AUTO_RECONCILE\`, and start-from-stale integration tests (backdate via \`sqlite3\` CLI with auto-skip when not installed).
- **Docs** — \`docs/v2-lifecycle.md\` gains a Reconciliation section; \`.env.example\` documents \`CODA_NO_AUTO_RECONCILE\`.

## Test results

- \`go test ./...\` — all green (6 packages).
- \`bash tests/run.sh\` — all green.
- \`bats test/\` — 17/17 green in \`v2-lifecycle.bats\`. 3 pre-existing failures on \`main\` (\`tmux-integration.bats\` env-dependent plus card120 edge case) are unrelated to this card; verified they fail identically on \`origin/main\` before any change.

## Brief-vs-reality gaps

- The brief §3.3 specified \`Reconcile\` returning \`(int, error)\` but §3.6 specified the CLI emitting \`[]ReconcileResult\` JSON. Those are incompatible; I chose \`([]ReconcileResult, error)\` so the CLI can forward structured results without a second DB query. The int count is recoverable as \`len(results)\`.
- The brief §8 called the schema bump \"v3\" in prose but the SQL in §3.1 inserts \`schema_version = 2\`. I followed the SQL (migration file \`002\` → version 2), which is also what \`internal/db/migrations\` expects (numbered files must be contiguous starting at 1).
- The brief §3.1's \`INSERT … SELECT\` column list did not exactly match the #156-shipped \`001_init.sql\` order (it listed \`session_id\` before \`tmux_session\`; the actual schema is the reverse). Migration uses the real column order.

## Scope discipline

Nothing touched outside \`internal/db/\`, \`internal/lifecycle/\`, \`internal/hooks/\`, \`cmd/coda-core/\`, \`docs/\`, \`.env.example\`, \`test/v2-lifecycle.bats\`. No new features beyond brief §3.

Closes #155.

### CODE REVIEW

**Review by evanstern — COMMENTED** [review]
## Architect review — PR #54 (card #155: v2 lifecycle reconciler)

Copilot caught five real issues — all valid, two are blockers. I have three additional architectural findings on top. **Hold for fixes before merge.**

---

### Copilot findings: confirmed

All five are real. Two are blockers in my read:

- **BLOCKER: FK constraint on DROP TABLE** (`002_reconciler.sql:28`) — same class of bug as the schema-guard ordering we fixed in #148. `DROP TABLE orchestrators` with `foreign_keys=ON` and a `features.orchestrator_id` FK will fail on any non-empty DB. Needs `PRAGMA defer_foreign_keys=ON` inside the transaction, or restructure to drop the FK reference before drop. Fix before merge.
- **BLOCKER: `tmux not found → (false, nil)`** (`reconciler.go:264`) — if tmux is not installed, every row with a `tmux_session` gets marked stale silently. `ErrNotFound` must return a non-nil error so Reconcile skips the transition. Fix before merge.
- **HIGH: `EPERM` treated as not-alive** (`reconciler.go:282`) — `EPERM` from `kill(pid, 0)` means the process exists, you lack permission. Marking it stale is wrong. Treat as alive.
- **HIGH: Non-optimistic UPDATE in `markOrchestratorStale`** (`reconciler.go:156`) — concurrent Start/Stop after the SELECT can be overwritten. Add `AND state IN ('starting','running','stopping') AND updated_at <= ?` to the WHERE clause.
- **HIGH: Same race in `markFeatureFailed`** (`reconciler.go:243`) — same fix needed.

---

### Architectural findings

**F1 (HIGH): `markFeatureFailed` fires no hook — asymmetry with `markOrchestratorStale`**

`markOrchestratorStale` fires `post-orchestrator-stale`. `markFeatureFailed` fires nothing. There is no comment explaining this omission. If intentional (no `post-feature-failed` hook defined yet), say so explicitly with a TODO pointing at the card that will add it. If it is an oversight, add the hook. Either way the asymmetry needs to be documented or closed.

**F2 (MEDIUM): `orchDeadReason` short-circuits on tmux-gone before checking pid**

For orchestrators launched as background processes that outlive their tmux session (e.g. a crashed session that left the pid running), tmux-gone returns a stale verdict even though the pid is alive. This is a false positive. The decision to treat tmux-gone-but-pid-alive as stale is architecturally defensible — but it needs to be a documented policy choice, not an implicit one. Add a comment in `orchDeadReason` stating the policy: tmux session is the primary liveness signal for orchestrators; a live pid without a tmux session is not a recoverable state.

**F3 (LOW): Docs do not call out that features without `tmux_session` are unreconcilable until #149**

The code correctly skips features with no `tmux_session`. The reconciler doc says "Features: tmux session only. `session_id` liveness is a message-bus (#149) concern." That covers the deferred case. It does not say what happens to features that never had a `tmux_session` written — they are permanently invisible to the reconciler. Add one sentence to the docs.

---

### Verdict

**Do not merge.** Two blockers (FK constraint, tmux-not-found), three highs (EPERM, optimistic update race x2), one architectural gap (hook asymmetry). The structural design is sound — `Prober` interface, freshness window, idempotency, stub-based tests are all correct. This is fixable in one more commit.

When the fixes land, ping me for round 2.

File: internal/lifecycle/reconciler.go [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+	}
+	err := syscall.Kill(pid, 0)
+	if err == nil {
+		return true, nil
+	}
+	if errors.Is(err, syscall.ESRCH) {
+		return false, nil
+	}
+	if errors.Is(err, syscall.EPERM) {
+		return false, nil
```
Comment (evanstern) [inline internal/lifecycle/reconciler.go]: **HIGH (Copilot confirmed): EPERM means alive, not dead.**

`syscall.EPERM` from `kill(pid, 0)` means the process exists but you lack permission to signal it. Returning `false, nil` here marks a live orchestrator stale. Treat EPERM as alive:

```go
if errors.Is(err, syscall.EPERM) {
    return true, nil
}
```

File: internal/lifecycle/reconciler.go [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+		return true, nil
+	}
+	var ee *exec.ExitError
+	if errors.As(err, &ee) {
+		return false, nil
+	}
+	if errors.Is(err, exec.ErrNotFound) {
+		return false, nil
+	}
+	return false, err
```
Comment (evanstern) [inline internal/lifecycle/reconciler.go:410]: **BLOCKER (Copilot confirmed): tmux not found → silent false, nil**

If tmux is not installed, `exec.ErrNotFound` returns `(false, nil)` — Reconcile interprets that as "session gone" and marks the row stale. Every orchestrator row with a tmux_session gets tombstoned silently.

Return an error instead:

```go
if errors.Is(err, exec.ErrNotFound) {
    return false, fmt.Errorf("tmux not found: %w", err)
}
```

**Review by evanstern — COMMENTED** [review]
Inline comments on specific findings — see thread.

File: internal/db/migrations/002_reconciler.sql [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+INSERT INTO orchestrators_new
+  (id, name, config_dir, state, tmux_session, session_id, port, pid,
+   started_at, stale_reason, created_at, updated_at)
+  SELECT id, name, config_dir, state, tmux_session, session_id, port, pid,
+         started_at, NULL, created_at, updated_at
+  FROM orchestrators;
+
+DROP TABLE orchestrators;
+ALTER TABLE orchestrators_new RENAME TO orchestrators;
```
Comment (evanstern) [inline internal/db/migrations/002_reconciler.sql:28]: **BLOCKER (Copilot confirmed): FK constraint will fail on DROP TABLE**

The DB opens with `foreign_keys=ON`. `features.orchestrator_id` has a FK referencing `orchestrators(id)`. Dropping `orchestrators` on a non-empty DB will fail with a foreign key constraint error.

Fix: add `PRAGMA defer_foreign_keys=ON;` inside the transaction boundary at the top of this migration. This defers FK checks until commit, by which point the renamed table satisfies the constraint.

File: internal/lifecycle/reconciler.go [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+		}
+	}
+	return ""
+}
+
+func (m *Manager) markOrchestratorStale(ctx context.Context, o *Orchestrator, reason string) error {
+	now := m.Now().Unix()
+	_, err := m.DB.ExecContext(ctx,
+		`UPDATE orchestrators SET state=?, stale_reason=?, updated_at=? WHERE id=?`,
+		StateStale, reason, now, o.ID)
```
Comment (evanstern) [inline internal/lifecycle/reconciler.go]: **HIGH (Copilot confirmed): Non-optimistic UPDATE races with concurrent Start/Stop**

If `StartOrchestrator` or `StopOrchestrator` transitions this row after the SELECT but before this UPDATE, the newer state gets overwritten to `stale`. Add a state guard:

```go
UPDATE orchestrators SET state=?, stale_reason=?, updated_at=?
WHERE id=? AND state IN ('starting','running','stopping') AND updated_at <= ?
```

Same fix needed in `markFeatureFailed` at line 243.

File: internal/lifecycle/reconciler.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+	if orchName != "" {
+		query += ` AND o.name = ?`
+		args = append(args, orchName)
+	}
+
+	rows, err := m.DB.QueryContext(ctx, query, args...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
```
Comment (evanstern) [inline internal/lifecycle/reconciler.go:206]: **HIGH: `markFeatureFailed` fires no hook — asymmetry with `markOrchestratorStale`**

`markOrchestratorStale` fires `post-orchestrator-stale`. `markFeatureFailed` fires nothing and has no comment explaining the omission.

If intentional (no `post-feature-failed` event defined yet), add a TODO:
```go
// TODO: fire post-feature-failed hook once the event is defined in internal/hooks
```
If it's an oversight, add the hook call before returning.

File: internal/lifecycle/reconciler.go [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+	p := m.probe()
+	var results []ReconcileResult
+	for _, o := range candidates {
+		reason := orchDeadReason(p, o)
+		if reason == "" {
+			continue
+		}
+		prev := o.State
+		if err := m.markOrchestratorStale(ctx, o, reason); err != nil {
+			return results, err
```
Comment (evanstern) [inline internal/lifecycle/reconciler.go:135]: **MEDIUM: `orchDeadReason` policy needs a comment**

This short-circuits on tmux-gone before checking pid, so a process that outlives its tmux session (pid alive, tmux gone) gets marked stale. That's a defensible policy choice, but it needs to be stated explicitly — otherwise the next reader will assume it's a bug.

Add a comment:
```go
// tmux session is the primary liveness signal for orchestrators.
// A live pid without a tmux session is not a recoverable state.
```

**Review by evanstern — COMMENTED** [review]
Inline architectural findings (supplement to summary review above).
